src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java

```diff
@@ -312,6 +391,205 @@ public void processElement(final ProcessContext context) throws Exception {
+
+            @Override
+            public DuplicationMetrics apply(Iterable<DuplicationMetrics> input) {
+                DuplicationMetrics metricsSum = new DuplicationMetrics();
+                int num = 0;
+                for (final DuplicationMetrics metrics : input) {
+                    if (metricsSum.LIBRARY == null) {
+                        metricsSum.LIBRARY = metrics.LIBRARY;
+                    }
+                    // This should never happen, as we grouped by key using library as the key.
+                    if (!metricsSum.LIBRARY.equals(metrics.LIBRARY)) {
+                        return null;
+                    }
+                    metricsSum.UNMAPPED_READS += metrics.UNMAPPED_READS;
+                    metricsSum.UNPAIRED_READS_EXAMINED += metrics.UNPAIRED_READS_EXAMINED;
+                    metricsSum.READ_PAIRS_EXAMINED += metrics.READ_PAIRS_EXAMINED;
+                    metricsSum.UNPAIRED_READ_DUPLICATES += metrics.UNPAIRED_READ_DUPLICATES;
+                    metricsSum.READ_PAIR_DUPLICATES += metrics.READ_PAIR_DUPLICATES;
+                    metricsSum.READ_PAIR_OPTICAL_DUPLICATES += metrics.READ_PAIR_OPTICAL_DUPLICATES;
+                }
+                metricsSum.PERCENT_DUPLICATION = 0.0;
+                metricsSum.ESTIMATED_LIBRARY_SIZE = 0L;
+                return metricsSum;
+            }
+        }
+
+        private PTransform<PCollection<? extends KV<String, DuplicationMetrics>>, PCollection<KV<String, DuplicationMetrics>>> finalizeMetrics() {
+            return ParDo
+                .named("finalize metrics")
+                .of(new DoFn<KV<String, DuplicationMetrics>, KV<String, DuplicationMetrics>>() {
+                        private static final long serialVersionUID = 1l;
+
+                        @Override
+                        public void processElement(final ProcessContext context) throws Exception {
+                            DuplicationMetrics metrics = context.element().getValue();
+                            metrics.READ_PAIRS_EXAMINED = metrics.READ_PAIRS_EXAMINED / 2;
+                            metrics.READ_PAIR_DUPLICATES = metrics.READ_PAIR_DUPLICATES / 2;
+                            metrics.READ_PAIR_OPTICAL_DUPLICATES = metrics.READ_PAIR_OPTICAL_DUPLICATES / 2;
+                            metrics.calculateDerivedMetrics();
+                            if (metrics.ESTIMATED_LIBRARY_SIZE == null) {
+                                metrics.ESTIMATED_LIBRARY_SIZE = 0L;
+                            }
+                            final KV<String, DuplicationMetrics> kv = KV.of(context.element().getKey(), metrics);
+                            context.output(kv);
+                        }
+                    });
+        }
+    }
+
+
+    /**
+     * Special coder for the DuplicationMetrics class.
+     */
+    private static final class DuplicationMetricsCoder extends CustomCoder<DuplicationMetrics> {
+        private static final long serialVersionUID = 1L;
```

Comment:
Well, I was given errors that it could not be encoded properly when I had no coder specified. Should I have just specified a default one? I don't recall the specific errors, but can check.